### ISSUE TITLE
nominal_value and actual_value

### ISSUE THREAD

**Issue by jnnr** [oemof/oemof-solph#547]
Flows have attributes `nominal_value` and `actual_value`. The latter, `actual_value`, gives information about the temporal profile of the flow. It is internally multiplied by `nominal_value`.

* Remark: A power timeseries does not necessary attain the maximal possible value or the nominal value in a given time interval. Scaling a timeseries so its maximum is not the same as scaling it to its general nominal value.
*  Terminology: Nennleistung is not necessarily the same as maximal power (https://www.energie-lexikon.info/leistung.html, section "Nennleistung"). The [docu](https://oemof.readthedocs.io/en/stable/oemof_solph.html#source-basic) does not draw this distinction . This means that a non-normalized power timeseries may contain values above the Nennleistung. E.g., PV can in some conditions produce more than its peak power which is measured under standard test conditions.

How to handle this?

My opinion: As flows are pretty flexible and can be used in many contexts, I suggest to avoid restricting its use just to meet the terminology of special cases. However, I would clearly state that `nominal_value` and `actual_value` are internally multiplied.

In this case, actual_value should be allowed to be >1. Without investment, this is already possible. In investment mode, it will lead to zero flow.

It seems to be caused by this constraint in blocks.py:

>
        def _max_investflow_rule(block, i, o, t):
            """Rule definition of constraint setting an upper bound of flow
            variable in investment case.
            """
            expr = (m.flow[i, o, t] <= (
                (m.flows[i, o].investment.existing + self.invest[i, o]) *
                 m.flows[i, o].max[t]))
            return expr
        self.max = Constraint(self.FLOWS, m.TIMESTEPS,
                              rule=_max_investflow_rule)

**Comment by simnh** [comment]
Generally I agree, but I do not see why the constraint above should lead to 0:

if `max` in some t = 100, there is no problem for the flow to be > 0 and < inf+ depending on the value of invest (leaving the existing investment out for now).

**Comment by jnnr** [comment]
I have to add that the case I am talking about is `fixed=True`. This leads to two constraints: `_investflow_fixed_rule` and `_max_investflow_rule`. 

Which means:
> 
flow = (invest + existing) * actual
flow <= (invest + existing) * max

where max seems to be set to 1 in the background. In the LP-file, this looks like this:

> 
c_e_InvestmentFlow_fixed(wind_electricity_17)_:
-1.089 InvestmentFlow_invest(wind_electricity)
+1 flow(wind_electricity_17)
= 0

c_u_InvestmentFlow_max(wind_electricity_17)_:
-1 InvestmentFlow_invest(wind_electricity)
+1 flow(wind_electricity_17)
<= 0

As soon as `actual value>0`, investment has to be 0.

**Comment by uvchik** [comment]
You are right. the reason is the default value of max. It makes things easier on the one hand but leads to problems if you do not use it as intended.

As we are heading a major release where API changes are possible you could change that.

I would just remove the default value. After that a lot of test will fail and you can step by step fix the code to make the test pass again.

**Comment by simnh** [comment]
this may be more complicated as it seems...but maybe its ok

**Comment by uvchik** [comment]
If `fixed=True` the max value should be ignored. In the non-investment case this is the case. In the investment case this is not the case. Maybe we should concentrate on that.

We should build the following constraint only if fixed is False:
`flow <= (invest + existing) * max`

**Comment by p-snft** [comment]
The name `actual_value` in flows is also misleading. I interpret `nominal_value` as "value the flow has by design (or data sheet)". In contrast, the `actual_value` would be "the value, the flow actually (meaning: really) has". So, it makes no sense to multiply the values. What the variable actually describes is more of a utilization (typically between 0 and 1).

**Comment by uvchik** [comment]
We could rename it to `fixed_value` and remove the `fixed` attribute because at the moment the `actual_value` is only used with `fixed=True`.

**Comment by uvchik** [comment]
Does anybody feel responsible for this issue?

**Comment by simnh** [comment]
I think renaming is a good idea.. But tree moving the fixed attribute not. :)

We could also do it like pypsa and append put (per unit) or something similar to indicate whether values are normalised or absolut.. Could also be

* actua value - normalised flow
* nominal value - nominal flow
*  absolut real flow - flow (new attribute)

I would also be fine with other names and or abbreviations. But something like the direction above I would like

**Comment by p-snft** [comment]
The `utilization` or `normalized_value` (I really dislike `actual_value`, which I would even consider a misnomer.) is actually a good thing to have for modeling. E.g. it can be utilized to have switchable flows by setting `flow.utilization[t] <= active[t]`.

For the `nominal_value`, I do not have a better suggestion for the name.

**Comment by uvchik** [comment]
I do not see a solution for the v0.3.0 release, so I postpone it to the next minor release.

We cannot break the API but we can introduce new parameter and define deprecated parameter.

**Comment by jakob-wo** [comment]
I am in favour of...
- renaming `actual_value` to `utilization` or `normalized_value` (I do prefere `utilization`).
- removing `fixed`, although I am not aware of the usage and the benifit that this attribute might have in other applications. If anyone is aware of a case where `fixed` is actually very useful or even essential, please share that knowledge.

**Comment by uvchik** [comment]
We have to clarify to different things.

1. Do we want to use the `actual_value` (or `utilization` or `normalized_value` in the future) for different purposes? At the moment it is only possible to fix the Flow-variable to this value multiplied with the `nominal_value`.

2. What should be a good name for this parameter.

In my opinion it will be easier to find a good name if we have answered question (1). 
Furthermore, I think that we need an additional parameter if the new name is generic. I could think of a name `fixed_value` (or similar) that makes it clear that the only purpose of this parameter is to fix the variable. If we name it `normalized_value` for various purposes than we still need a parameter `fixed` or `fixed_to_normalised_value`.

**Comment by p-snft** [comment]
I do not see the `actual value` to be used for different purposes and would vote for renaming it to `utilization`. Also (looking at #631), I believe that setting the `utilization` should automatically make it fixed.

PS: Other possible options would require to redesign `solph` so that it would directly use `actual_value` without multiplying it by `nominal_value`.

**Comment by uvchik** [comment]
Did you find `utilisation` in other models? In my head it is not connected to a fixed variable. For me it would be more intuitive to use a more speaking name from the model perspective: `max`, `min` and `fix` or `set`. But naming discussions are always annoying and I would follow the majority. 

It would be possible to introduce an absolute fix value. A value that shouldn't be used together with the investment mode. But as it breaks the existing concept a little bit and you should have good reasons for that :smirk:

**Comment by p-snft** [comment]
Actually, `min`, `max` and `actual_value` are of similar type and should be called similarly. In this spirit, I like both, `fix` and `set` as parameters when constructing the model, but the name should stay the same when it is part of the result. What's about `load`, `load_max`, and `load_min`?

What I found in other models is just "value" without any further description. But these do also just take the values as they are -- without `nominal_value`. In these cases also `min` and `max` are given in absolute values.

**Comment by simnh** [comment]
we can remove the fixed attribute. However `utilization` is kind of too general for me. I would stick with normalized value. But you can decide as you like. From my perspective it could also stay as it is. If you like to change it, I am fine with any change. :-)

**Comment by jnnr** [comment]
I am in favor of

* Removing the `fixed` attribute. When `actual_value` is passed, it should be fixed.
* Renaming `actual_value`, but keep the multiplication with nominal value. 

Regarding names, `utilisation` (German: Auslastung) seems a good hint to the fact that normalized timeseries are expected.

Another term would be `capacity_factor` (The net capacity factor is the unitless ratio of an actual electrical energy output over a given period of time to the maximum possible electrical energy output over that period. https://en.wikipedia.org/wiki/Capacity_factor). The timeseries passed are hourly (or timeincrement respective) capacity factors.

I made an overview over the suggestions that were already made. If the table still misses something or some combination or I got something wrong, please add them. In the end, when all good ideas are included we can do a vote.

| `nominal_value` | `actual_value` | `min` | `max` |
| ------------- | ------------- | ------------- | ------------- |
| `normalized_flow`  |  `nominal_flow` |  |
| `nominal_value`  | `fix`  | `min` | `max` |
| `nominal_value`  | `load`  | `load_min` | `load_max` |
| `nominal_value`  | `fixed_value`  | `min_value` | `max_value` |
| `nominal_value`  | `utilisation`  | `min` | `max` |
| `nominal_value`  | `capacity_factor`  | `min` | `max` |

**Comment by jnnr** [comment]
I prefer [2] :)

**Comment by p-snft** [comment]
Note that we do also have the internal variable ´flow´, e.g.:

```Python
def _minimum_flow_rule(block, i, o, t):
    """Rule definition for MILP minimum flow constraints.
    """
    expr = (self.status[i, o, t] *
        m.flows[i, o].min[t] * m.flows[i, o].nominal_value <=
        m.flow[i, o, t])
```

So, "fix" would not be the name of optimized values. (Else, I would have preferred something else.) Further, it has some similarity to min/max, which are as well normalized values. In conclusion, [2] is a good option.

BTW: `Flow.flow` seems to contain what I would understand when reading about the actual value of a flow.

**Comment by jakob-wo** [comment]
>  it has some similarity to min/max, which are as well normalized values.

That makes [2] a convincing option to me and gives us a coherent naming.

**Comment by jnnr** [comment]
I interpreted the feedback as votes, counted them and added them to the table.

**Comment by FranziPl** [comment]
I'm totally fine with [2].

I would like to ask again about the second point in this issue, besides the renaming. We were talking about the possibility, that actual_value (or in future maybe fix) can be bigger than 1 in investment mode (which was not possible).  If I understood right, the idea is, to remove the fixed attribute, but not change the constraints related to it. So this problem would be not adressed.
Uwe made a suggestion for this on 11 Dec 2018:
> 
> If `fixed=True` the max value should be ignored. In the non-investment case this is the case. In the investment case this is not the case. Maybe we should concentrate on that.
> 
> We should build the following constraint only if fixed is False:
> `flow <= (invest + existing) * max`

So I would appreciate the possibility of fix-values bigger one, but I don't feel able to work on it.

**Comment by p-snft** [comment]
From my perspective, it does not make sense to define a fixed value greater than the maximum value. So, the default value for max (=1) is a problem here.

Anyway, the constraints for min and max should not be created if the value is fixed, also from a performance perspective. There is nothing left to optimize.

**Comment by uvchik** [comment]
@jnnr will you open a PR to

1. rename the `actual_value`, remove the `fixed` value
2. to make sure min/man constraints are not build for the investment flow if `fix is not None`

I may help to fix the second aspect.

**Comment by jnnr** [comment]
Yes, I think we are approaching a PR :). I guess we also need

    3. Deprecation warning for fixed.

I am happy to do it, however not before November, because of other pressing deadlines.

**Comment by uvchik** [comment]
...and a `DeprecateWarning` for the renaming.

We may want to use a `FutureWarning` because as I know it a `DeprecateWarning` is suppressed by default and therefore may not be noticed by the users but we can check this.

**Comment by p-snft** [comment]
As this changes the API, it should be in v0.4

**Comment by uvchik** [comment]
Closed by #680
